bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttService.java

```diff
@@ -29,211 +34,315 @@
+            brokerConnections.remove(connection.getName(), connection);
+            for (MqttBrokersObserver o : brokersObservers) {
+                o.brokerRemoved(connection);
             }
-            return conn;
         }
     }
 
     /**
+     * Remove a broker connection by name
+     *
+     * @param brokerName The broker name
+     */
+    public void removeBrokerConnection(String brokerName) {
+        synchronized (brokerConnections) {
+            MqttBrokerConnection connection = brokerConnections.get(brokerName.toLowerCase());
+            if (connection == null) {
+                return;
+            }
+            brokerConnections.remove(brokerName);
+            for (MqttBrokersObserver o : brokersObservers) {
+                o.brokerRemoved(connection);
+            }
+        }
+    }
+
+    public Collection<MqttBrokerConnection> getAllBrokerConnections() {
+        return brokerConnections.values();
+    }
+
+    /**
      * Register a new connection observer that could act on MQTT connection changes.
+     * This is deprecated, please register on the broker connection object instead.
      *
+     * @deprecated
      * @param brokerName Name of the broker that connection should be observed.
      * @param connectionObserver The connection observer that should be informed about connection changes.
      */
+    @Deprecated
     public void registerConnectionObserver(String brokerName, MqttConnectionObserver connectionObserver) {
-        getConnection(brokerName).addConnectionObserver(connectionObserver);
+        MqttBrokerConnection brokerConnection = getBrokerConnection(brokerName);
+        if (brokerConnection != null) {
+            brokerConnection.addConnectionObserver(connectionObserver);
+        }
     }
 
     /**
      * Unregister an existing connection observer.
      *
+     * @deprecated
      * @param brokerName Name of the broker that connection has been observed.
      * @param connectionObserver The connection observer that should not be informed anymore.
      */
+    @Deprecated
     public void unregisterConnectionObserver(String brokerName, MqttConnectionObserver connectionObserver) {
-        getConnection(brokerName).removeConnectionObserver(connectionObserver);
+        MqttBrokerConnection brokerConnection = getBrokerConnection(brokerName);
+        if (brokerConnection != null) {
+            brokerConnection.removeConnectionObserver(connectionObserver);
+        }
     }
 
     /**
      * Register a new message consumer which can process messages received on
      *
+     * @deprecated
      * @param brokerName Name of the broker on which to listen for messages.
      * @param mqttMessageConsumer Consumer which will process any received message.
      */
+    @Deprecated
     public void registerMessageConsumer(String brokerName, MqttMessageConsumer mqttMessageConsumer) {
-
         mqttMessageConsumer.setEventPublisher(eventPublisher);
```

Comment:
not part of your PR and deprecated but I guess this should be done in line 283 only when the consumer was added to the connection.